File: Python/openstk/platforms/opengl/gfx/openglopenengine.py
from __future__ import annotations
import os, re
from numpy import ndarray
from quaternion import quaternion
from OpenGL.GL import *
from openstk.core import _throw, AsyncCoroutineQueue, CellManager, CellBuilder
from openstk.gfx import Shader

class OpenGLOpenEngine:
    desiredWorkTimePerFrame: float = 1.0 / 200
    def __init__(self, manager: callable, sunCycle: bool = False):
        if not manager: raise Exception('manager')
        self.queue: AsyncCoroutineQueue = AsyncCoroutineQueue()
        self.cellManager: CellManager = manager(self.queue) or _throw('manager')
        self.query = self.cellManager.query
        self.world: int = 0
        self._cell: ICell = None
        #self.playerTransform: Transform
        self.camera: object = None

    def dispose(self) -> None: pass

    @property
    def cell(self) -> ICell: return self._cell
    @cell.setter
    def cell(self, value: ICell) -> None:
        if self._cell == value: return
        self._cell = value
        if self.cellChanged: self.cellChanged(self._cell)

    cellChanged: callable = None

    async def update(self) -> None:
        # The current cell can be null if the player is outside of the defined game world.
        if self.camera and (not self._cell or not self._cell.isInterior): await self.cellManager.updateCells(self.camera.location)
        await self.queue.run(OpenGLOpenEngine.desiredWorkTimePerFrame)

    def _createPlayer(self, position: Vector3, rotation: quaternion) -> None:
        self.camera = position

    # Spawns the player inside using the cell's grid coordinates.
    async def spawnPlayer(self, db: ICellDatabase):
        self.cell = self.query.findCell(db.cellId)
        assert(self._cell)
        self._createPlayer(db.playerPosition, db.playerRotation)
        cell = self.cellManager.beginCell(db.cellId)
        await self.queue.waitFor(cell.task)

File: dotnet/Platform/OpenStack.OpenGL/Gfx/OpenGLOpenEngine.cs
using OpenStack.Gfx.Egin;
using System;
using System.Collections;
using System.Diagnostics;
using System.Numerics;
using static OpenStack.CellManager;

namespace OpenStack.Gfx.OpenGL;

public class OpenGLOpenEngine : IDisposable {
    const float DesiredWorkTimePerFrame = 1.0f / 200;
    readonly IQuery Query;
    readonly CellManager CellManager;
    readonly CoroutineQueue Queue = new();
    protected int World;
    //protected Transform PlayerTransform;
    public Camera Camera;

    public OpenGLOpenEngine(Func<CoroutineQueue, CellManager> manager, bool sunCycle = false) {
        if (manager == null) throw new ArgumentNullException(nameof(manager));
        CellManager = manager(Queue) ?? throw new ArgumentNullException(nameof(manager));
        Query = CellManager.Query;
    }

    public void Dispose() { } // Query.Dispose();

    ICell _cell;
    public ICell Cell {
        get => _cell;
        private set {
            if (_cell == value) return;
            _cell = value;
            CellChanged?.Invoke(_cell);
        }
    }

    public event Action<ICell> CellChanged;

    public virtual void Update() {
        // The current cell can be null if the player is outside of the defined game world.
        if (Camera != null && (_cell == null || !_cell.IsInterior)) CellManager.UpdateCells(Camera.Location);
        Queue.Run(DesiredWorkTimePerFrame);
    }

    void CreatePlayer(Vector3 position, Quaternion roation) {
    }

    /// <summary>
    /// Spawns the player inside using the cell's grid coordinates.
    /// </summary>
    /// <param name="position">The target position of the player.</param>
    /// <param name="roation">The target position of the player.</param>
    public void SpawnPlayer(ICellDatabase db) {
        Cell = Query.FindCell(db.CellId);
        Debug.Assert(Cell != null);
        CreatePlayer(db.PlayerPosition, db.PlayerRotation);
        var cell = CellManager.BeginCell(db.CellId);
        Queue.WaitFor(cell.Task);
    }
}
